自動更新機能の実装

yoshikyu opened the pull request:
What
Chatspaceの自動更新機能の実装

Why
UX向上の為

yoshikyu committed:
自動更新実装

TOMARUMARU reviewed (commented):
いまだとcontrollerの時点で@group.messagesで全てのmessageを取得してしまっているので、
現在doneの中でやっている
message.id > id
の処理をcontrollerで絞ってしまいましょう！

yoshikyu committed:
非同期メッセージ取得方法の変更実装

・元々はデータベースのコメントをすべて取得し、jsファイルにて一番最後のメッセージのidと画面上のidを比較する、という方法で実装していた。
・今回はajaxで画面上の最後のメッセージのidを送り、その後にデータベース上で一番最後のメッセージのidと比較、新しいものがあれば非同期で送る
、という実装に変更した。

yoshikyu committed:
調整

yoshikyu committed:
調整

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,52 @@
+        dataType:'json',
+        data: {id: lastId}
+  	  })
+
```
$.ajax({}).done({}).fail({}); でひとまとまりです。 .done や .fail の前に空行は避けましょう。 連続すべきです。

MrUrakou commented on app/views/messages/_message.html.haml:
```diff
@@ -1,4 +1,4 @@
-.message.clearfix 
+.message.clearfix{"data-message-id": "#{message.id}"}
```
{ 'data-message-id': message.id }としましょう。

MrUrakou commented on app/controllers/messages_controller.rb:
```diff
@@ -4,7 +4,11 @@ class MessagesController < ApplicationController
     @message = Message.new
     @messages = @group.messages.includes(:user)
-    @users = User.all
+    @new_messages = @messages.where("id>#{params[:id]}")
```
format.jsonに続けて{}で記述すればワンライナーにできます。

あと("id>#{params[:id]}")の記述って合ってるんですかね？
参考url
https://railsguides.jp/active_record_querying.html
こちらの「2 条件」あたりを参考に。

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,52 @@
+            {$('.messages').append(insertHTML);
+            };
+        });
+        $('.messages').animate({scrollTop: $('.messages')[0].scrollHeight},4500,'linear');
```
スクロール機能は関数化して呼び出しましょう。

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,52 @@
+  function interval(){
+    setInterval(function(){
+      var lastId = $('.message:last').data('messageId');
+      console.log(lastId);
```
デバック消しましょう。

yoshikyu committed:
修正

yoshikyu committed:
調整

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,51 @@
+    return html
+  };
+  function scrollDown(){
+    $('.messages').animate({scrollTop: $('.messages')[0].scrollHeight},4500,'linear' );
```
,の後は半角スペースです。

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,51 @@
+      })
+    }, 5000 );
+  };
+  if (window.location.href.match(/\/groups\/\d+\/messages/)){
```
hrefだとデプロイしてpathが変わると動かないのでpathnameとしましょう。

MrUrakou commented on app/controllers/messages_controller.rb:
```diff
@@ -4,7 +4,10 @@ class MessagesController < ApplicationController
-    @users = User.all
+    respond_to do |format|
+      format.html
+      format.json{@new_messages = @messages.where("id>?", params[:id])}
```
半角空けたいですね。
"id > ?"

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,51 @@
+  }
+  function interval(){
+    setInterval(function(){
+      var lastId = $('.message:last').data('messageId');
```
viewと紐づいているのでdata('message-id')の気がしますがいかがでしょうか？

MrUrakou commented on app/assets/javascripts/group.js:
```diff
@@ -0,0 +1,51 @@
+          <p class="bottom--message">
+            ${ message.content }
+            ${ message.image?
+              `<img src = '${ message.image }', width="200", height="150", class='lower-message__image'>` :''}
```
width="200", height="150"
なるべくならcssに記述したいです。

yoshikyu committed:
修正

yoshikyu committed:
調整
